src/test/java/plugins/GitPluginTest.java

```diff
@@ -209,16 +202,226 @@ public void update_submodules_recursively() throws IOException, InterruptedExcep
+                .sparseCheckout().addPath(SUB_DIR);
+
+        job.addShellStep("test ! -f " + TEST_FILE + " && test -f " + SUB_DIR + "/" + TEST_FILE);
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+    }
+
+    @Test
+    public void ancestry_strategy_to_choose_build() {
+        final String TEST_BRANCH = "testBranch";
+
+        GitRepo repo = buildGitRepo();
+        repo.createBranch(TEST_BRANCH);
+        repo.changeAndCommitFoo("Commit1 on master");
+        repo.checkout(TEST_BRANCH);
+        repo.changeAndCommitFoo("commit1 on " + TEST_BRANCH);
+        String sha1 = repo.getLastSha1();
+        repo.changeAndCommitFoo("commit2 on " + TEST_BRANCH);
+
+        repo.transferToDockerContainer(host, port);
+
+        job.useScm(GitScm.class)
+                .url(repoUrl)
+                .branch("")
+                .credentials(USERNAME)
+                .chooseBuildStrategy("Ancestry", 1, sha1);
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+        // TODO Multiple selected branches create multiple builds, these should also be verified
+
+        assertThat(
+                b.getConsole(),
+                Matchers.containsRegexp(
+                        "Checking out Revision .* \\(origin/"+TEST_BRANCH+"\\)",
+                        Pattern.MULTILINE
+                )
+        );
+    }
+
+    @Test
+    public void inverse_strategy_to_choose_build() {
+        final String BRANCH_NAME = "secondBranch";
+
+        GitRepo repo = buildGitRepo();
+        repo.createBranch(BRANCH_NAME);
+        repo.transferToDockerContainer(host, port);
+
+        job.useScm(GitScm.class)
+                .url(repoUrl)
+                .credentials(USERNAME)
+                .chooseBuildStrategy("Inverse");
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+
+        assertThat(
+                b.getConsole(),
+                Matchers.containsRegexp(
+                        "Checking out Revision .* \\(origin/"+BRANCH_NAME+"\\)",
+                        Pattern.MULTILINE
+                )
+        );
+    }
+
+    ////////////////////
+    // HELPER METHODS //
+    ////////////////////
+
     private String getRevisionFromConsole(String console) {
         Pattern p = Pattern.compile("(?<=\\bRevision\\s)(\\w+)");
         Matcher m = p.matcher(console);
         assertThat(m.find(), is(true));
         return m.group(0);
     }
 
-    private GitRepo buildGitRepo() throws IOException, InterruptedException {
+    private GitRepo buildGitRepo() {
         GitRepo repo = new GitRepo();
         repo.changeAndCommitFoo("Initial commit");
         return repo;
     }
+
+    /**
+     * Invoked by {@link #clean_after_checkout()} and {@link #clean_before_checkout()}
+     *
+     * @param before Select "clean before" or "clean after"
+     */
+    private void test_clean_while_checkout(boolean before) {
```

Comment:
@uhafner, the only behavioral difference between them is the order of operations.  The checkout happens and is either preceded or followed by a git clean.

The rationale for adding "clean before" was that there are cases where a git checkout will fail if not preceded by a clean.  If the git checkout fails, then the "clean after" would not be run.  Thus, "clean before" was added in an attempt to resolve those cases where "clean after" was not reached because checkout failed.

Refer to [JENKINS-22510](https://issues.jenkins-ci.org/browse/JENKINS-22510) for more discussion on the issue.

The git plugin still has several cases where a checkout does not construct into an existing workspace does not create the same directory structure as a checkout into an empty workspace.  Refer to [JENKINS-38860](https://issues.jenkins-ci.org/browse/JENKINS-38860), [JENKINS-41553](https://issues.jenkins-ci.org/browse/JENKINS-41553), and [JENKINS-43977](https://issues.jenkins-ci.org/browse/JENKINS-43977)